Extensions/Security/Authentication.md

```diff
@@ -0,0 +1,35 @@
+# Authentication Extension
+
+_This extension specification is currently incubating.  While incubating the version is 0._
+
+## Introduction
+Authentication is a necessary component to any real world application. This extension specification provides a standardized mechanism for including both the type of credentials and the credentials in metadata payloads.
+
+## Metadata Payload
+This metadata type can be used in a per connection or per stream, and not individual payloads and as such it **MUST** only be used in frame types used to initiate interactions and payloads.  This includes [`SETUP`][s], [`REQUEST_FNF`][rf], [`REQUEST_RESPONSE`][rr], [`REQUEST_STREAM`][rs], [`REQUEST_CHANNEL`][rc], and [`PAYLOAD`][p].  The Metadata MIME Type is `message/x.rsocket.authentication.v0`.
+
+[s]:  ../../Protocol.md#frame-setup
+[p]:  ../../Protocol.md#frame-payload
+[rc]: ../../Protocol.md#frame-request-channel
+[rf]: ../../Protocol.md#frame-fnf
+[rr]: ../../Protocol.md#frame-request-response
+[rs]: ../../Protocol.md#frame-request-stream
+
+
+### Metadata Contents
+```
+     0                   1                   2                   3
+     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    |A| Auth ID/Len |   Authentication Type                        ...
+    +---------------+---------------+---------------+---------------+
+    |                     Authentication Payload                   ...
+    +---------------+-----------------------------------------------+
+```
+
+* (**A**)uthentication Type: Authentication type is a well known value represented by a unique integer.
```

Comment:
It's not clear which bit value does what.
Maybe we should be more explicit, i.e. `0 means Auth Length, 1 means Auth Id`.

Reply:
@stevegury Thanks for the feedback. I tried to follow the conventions I found in the existing specs. The clarification is on the next line just as it is done in [Composite Metadata](https://github.com/rsocket/rsocket/blob/7278d1dd7963893be77069bb3b699489d65f42ac/Extensions/CompositeMetadata.md#metadata-contents):

> If A flag is set, indicates a [Well-known Auth Type ID][wk].  If A flag is not set, indicates the Authentication Type Length in bytes.

I pushed a fix in https://github.com/rsocket/rsocket/pull/275/commits/de71eaac27ee5b9d231515a9662eed020c8b09a3 I kept the commit separate, so it was easier to see the changes. I can clean up the commits later if desired.

If we decide to stick with these changes, I'd suggest we also update Composite Metadata.